torch/csrc/jit/tensorexpr/ir_serializer.cpp

```diff
@@ -0,0 +1,557 @@
+    buffer_mapping_json[mapping.first] = visitAndSerialize(mapping.second);
+  }
+  json loop_options_json = {
+      {"buffer_mapping", buffer_mapping_json},
+      {"gpu_block_index", v->loop_options().gpu_block_index()},
+      {"gpu_thread_index", v->loop_options().gpu_thread_index()},
+  };
+  json for_json = {
+      {"stmt_type", "For"},
+      {"var", visitAndSerialize(v->var())},
+      {"start", visitAndSerialize(v->start())},
+      {"stop", visitAndSerialize(v->stop())},
+      {"loop_options", loop_options_json},
+  };
+  if (v->body()) {
+    for_json["body"] = visitAndSerialize(v->body());
+  }
+  putJSON(v, for_json);
+}
+
+void IRSerializer::visit(const Broadcast* v) {
+  JSON_CACHE_GUARD();
+  putJSON(
+      v,
+      {
+          {"expr_type", "Broadcast"},
+          {"value", visitAndSerialize(v->value())},
+          {"lanes", v->lanes()},
+          {"dtype", serializeDtype(v->dtype())},
+      });
+}
+
+void IRSerializer::visit(const IfThenElse* v) {
+  JSON_CACHE_GUARD();
+  putJSON(
+      v,
+      {
+          {"expr_type", "IfThenElse"},
+          {"true_value", visitAndSerialize(v->true_value())},
+          {"false_value", visitAndSerialize(v->false_value())},
+          {"condition", visitAndSerialize(v->condition())},
+          {"dtype", serializeDtype(v->dtype())},
+      });
+}
+
+void IRSerializer::visit(const BaseCallNode* v) {
+  // should have visited the inherited class
+  assert(false);
+}
+
+void IRSerializer::visit(const Intrinsics* v) {
+  putJSON(
+      v,
+      {{"expr_type", "Intrinsics"},
+       {"dtype", serializeDtype(v->dtype())},
+       {"func_name", v->func_name()},
+       {"params", visitAndSerialize(v->params())}});
+}
+
+void IRSerializer::visit(const Allocate* v) {
+  JSON_CACHE_GUARD();
+  // TODO: why is allocate a stmt and not an expr ?
+  putJSON(
+      v,
+      {
+          {"stmt_type", "Allocate"},
+          {"dtype", serializeDtype(v->dtype())},
+          {"buffer_var", visitAndSerialize(v->buffer_var())},
+          {"dims", visitAndSerialize(v->dims())},
+      });
+}
+
+void IRSerializer::visit(const Free* v) {
+  JSON_CACHE_GUARD();
+  putJSON(
+      v,
+      {
+          {"stmt_type", "Free"},
+          {"buffer_var", visitAndSerialize(v->buffer_var())},
+      });
+}
+
+void IRSerializer::visit(const Buf* v) {
+  JSON_CACHE_GUARD();
+  putJSON(
+      v,
+      {{"expr_type", "Buf"},
+       {"dtype", serializeDtype(v->dtype())},
+       {"base_handle", visitAndSerialize(v->base_handle())},
+       {"dims", visitAndSerialize(v->dims())}});
+}
+
+void IRSerializer::visit(const Cond* v) {
+  JSON_CACHE_GUARD();
+  putJSON(
+      v,
+      {{"stmt_type", "Cond"},
+       {"condition", visitAndSerialize(v->condition())},
+       {"true_stmt", visitAndSerialize(v->true_stmt())},
+       {"false_stmt", visitAndSerialize(v->false_stmt())}});
+}
+
+void IRSerializer::visit(const Term* v) {
+  JSON_CACHE_GUARD();
+  // TODO: why is Term not a a kernel scoped object ?
```

Comment:
This seems like a bug, it should be cleaned up at scope exit. Ideally it should be cleaned up at the end of simplification, but I don't think we have hierarchical kernel scope arenas yet.

Term and Polynomial (and MaxTerm and MinTerm) are temporary nodes that should exist only during arithmetic simplification, so it doesn't seem important to include them here.